### PR TITLE
Add extension tutorial

### PR DESCRIPTION
**Pull request by mitmul**
This PR adds a tutorial of how to create your own trainer extension.
Related to https://github.com/chainer/chainer/issues/2576

TODO:

- [x] Finish the chapter of @make_extension decorator
- [x] Write a chapter for a class inherited from training.Extension

### CODE REVIEW

File: docs/source/tutorial/extension.rst [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    * :ref:`by defining a simple function<function>`
+    * :ref:`by defining a function decorated with @make_extension <decorator>`
+    * :ref:`by defining a class inherited from Extension class <class>`
+
+.. _basic:
+
+What is trainer Extension?
+--------------------------
+
+:class:`~chainer.training.Extension` is a callable object that takes a :class:`~chainer.training.Trainer` object as an argument. Adding an :class:`~chainer.training.Extension` to a :class:`~chainer.training.Trainer` using :meth:`~chainer.training.Trainer.extend` method, the :class:`~chainer.training.Extension` will be called at given timing you specified by using ``trigger`` object (See the details in :ref:`trigger`)
```
[the remark this answers is not in the document]
Comment (mitmul) [reply docs/source/tutorial/extension.rst]: Thank you! fixed

File: docs/source/tutorial/extension.rst [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    def lr_drop(trainer):
+        trainer.updater.get_optimizer('main').lr *= 0.1
+
+Then you can add this function to a :class:`~chainer.training.Trainer` object via :meth:`~chainer.training.Trainer.extend` method.
+
+.. code-block:: python
+
+    trainer.extend(lr_drop, trigger=(10, 'epoch'))
+
+It performs learning rate dropping at every 10 epochs by multiplying 0.1 with the current learning rate.
```
[the remark this answers is not in the document]
Comment (mitmul) [reply docs/source/tutorial/extension.rst]: fixed

File: docs/source/tutorial/extension.rst [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+:meth:`~chainer.training.make_extension` is a decorator that adds some attributes to a given function. For example, the simple extension we created above can be written in this form:
+
+.. testcode::
+
+    @make_extension(trigger=(10, 'epoch'))
+    def lr_drop(trainer):
+        trainer.updater.get_optimizer('main').lr *= 0.1
+
+The difference between the above one and this is whether it has a default ``trigger`` or not. In this latter case, :meth:`lr_drop` has its default ``trigger`` so that unless another ``trigger`` is specified via :meth:`~chainer.training.Trainer.extend` method, the ``trigger`` specified in :meth:`~chainer.training.make_extension` is used as default. So the code below acts the same as the former example, i.e., it reduces the learning rate at every 10 epochs.
```
[the remark this answers is not in the document]
Comment (mitmul) [reply docs/source/tutorial/extension.rst]: fixed

File: docs/source/tutorial/extension.rst [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+``trigger`` can also be given to the :meth:`~chainer.training.trainer.extend` method that adds an extension to a :class:`~chainer.training.Trainer` object. The priority of ``trigger``\ s is as follows:
+
+- When both :meth:`~chainer.training.Trainer.extend` and a given :class:`~chainer.training.Extension` have ``trigger``\ s, the ``trigger`` given to :meth:`~chainer.training.Trainer.extend` is used.
+- When ``None`` is given to :meth:`~chainer.training.Trainer.extend` as the ``trigger`` argument and a given :class:`~chainer.training.Extension` has ``trigger``, the ``trigger`` given to the :class:`~chainer.training.Extension` is used.
+- When both ``trigger`` attributes in :meth:`~chainer.training.Trainer.extend` and :class:`~chainer.training.Extension` are ``None``, the :class:`~chainer.training.Extension` will be fired at every iteration.
```
[the remark this answers is not in the document]
Comment (mitmul) [reply docs/source/tutorial/extension.rst]: fixed

**Comment by mitmul** [comment]
@bkvogel Thank you for the review. I resolved all points you pointed out, so please check this PR again and merge! :)

**Comment by bkvogel** [comment]
Sorry for the delay in approving. LGTM!

**Comment by niboshi** [comment]
Should this PR be backported?

**Pull request merged**